### PR TITLE
Remove dependency of acceptance tests to other operators

### PR DESCRIPTION
**Pull request by pratikjagrut**
### Motivation

Currently, there is a number of scenarios that depends on other operators (such as the PostgreSQL, etcd or knative operators). That makes the acceptance tests heavy, dependent on those operators and more difficult to be executed without a cluster-admin role.

Also, the features are organized in a way that in some cases a scenario depends on the PostgreSQL operator, even if it does not actually use it - the dependency is in the {{Background: }} section of a feature file where the above-mentioned scenarios are placed.

### Changes

Segregating the external dependent operator scenarios and annotating them with @examples tag.

### CODE REVIEW

**Comment by Avni-Sharma** [comment]
For the Remove dependency of acceptance tests to other operators how are we actually checking whether the time elapsed would be less in this case than what actually exists in the master? Are we checking it in the CI? How are we confirming that removing all dependencies will indeed result in a speedy test pass/fail scenario? As mentioned in https://issues.redhat.com/browse/APPSVC-812?focusedCommentId=15736777&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-15736777 the expectation is also to see at the end of the work less time spent in developers's inner loop tests , in addition to ofcourse the already existing acceptance criteria.

File: test/acceptance/features/customEnvVar.feature [hunk 1]
```diff
@@ -7,11 +7,11 @@ Feature: Inject custom env variable into application
     Background:
         Given Namespace [TEST_NAMESPACE] is used
         * Service Binding Operator is running
+        * OLM Operator "backend" is running
```
Comment (Avni-Sharma) [inline test/acceptance/features/customEnvVar.feature:10]: If we have OLM operator as backend in the Background then like previous feature files we can technically have olm as a tag to the entire feature file itself.

File: test/acceptance/features/deleteBackendServiceAndRecreate.feature [hunk 2]
```diff
@@ -6,11 +6,11 @@ Feature: Reconcile when BackingService CR got deleted and recreated
     Background:
         Given Namespace [TEST_NAMESPACE] is used
         * Service Binding Operator is running
+        * OLM Operator "backend" is running
```
Comment (Avni-Sharma) [inline test/acceptance/features/deleteBackendServiceAndRecreate.feature:9]: Refer https://github.com/redhat-developer/service-binding-operator/pull/880#discussion_r584524746

File: test/acceptance/features/bindAppToService.feature [hunk 3]
```diff
@@ -7,183 +7,256 @@ Feature: Bind an application to a service
...
+                data:
+                    dbCredentials: backend-secret
+            """
         When Service Binding is applied
             """
             apiVersion: binding.operators.coreos.com/v1alpha1
             kind: ServiceBinding
             metadata:
-                name: binding-request-a-d-s
+                name: service-binding-a-b-s
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:62]: is this change really needed in this PR?

File: test/acceptance/features/bindAppToService.feature [hunk 4]
```diff
@@ -7,183 +7,256 @@ Feature: Bind an application to a service
...
+        * The Secret is present
+            """
+            apiVersion: v1
+            kind: Secret
+            metadata:
+                name: backend-secret
+            stringData:
+                username: AzureDiamond
+                password: hunter2
+            """
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:42]: does each scenario in this feature require existence of this secret?

File: test/acceptance/features/bindAppToService.feature [hunk 5]
```diff
@@ -7,183 +7,256 @@ Feature: Bind an application to a service
...
+
+    Scenario:  Bind an application to backend service in the following order: Application, Service Binding and backend service
+        Given Generic test application "gen-app-a-s-b" is running
+        And Service Binding is applied
             """
             apiVersion: binding.operators.coreos.com/v1alpha1
             kind: ServiceBinding
             metadata:
-                name: binding-request-a-s-d
+                name: service-binding-a-s-b
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:86]: same as above

File: test/acceptance/features/postgresqlOperatorExamples.feature [hunk 6]
```diff
@@ -0,0 +1,331 @@
+@examples
+Feature: Bind and application to postgresql database
```
Comment (pedjak) [inline test/acceptance/features/postgresqlOperatorExamples.feature:2]: type: `and`

File: test/acceptance/features/bindAppToMultipleServices.feature [hunk 7]
```diff
@@ -7,9 +7,10 @@ Feature: Bind a single application to multiple services
     Background:
         Given Namespace [TEST_NAMESPACE] is used
         * Service Binding Operator is running
+        * OLM Operator "backend" is running
```
Comment (pedjak) [inline test/acceptance/features/bindAppToMultipleServices.feature:10]: is this fake operator really required for the scenarios in this feature file to pass? My personal feeling is no - what most of the scenarios need is just existence of appropriate CRD so that service instances can be created. We should really identify scenarios where operators ClusterServiceVersion is needed and only then use this step. For all other, we should define another step, e.g. `Operator X is running` where we just deploy needed CRDs.

File: test/acceptance/features/bindAppToService.feature [hunk 8]
```diff
@@ -7,183 +7,256 @@ Feature: Bind an application to a service
     Background:
         Given Namespace [TEST_NAMESPACE] is used
         * Service Binding Operator is running
-        * PostgreSQL DB operator is installed
+        * OLM Operator "backend" is running
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:10]: is this step needed at given that in the next step we apply right CRD?

**Review by pedjak — CHANGES_REQUESTED** [review]
a number of `@olm` tags can be removed now.

File: test/acceptance/features/bindAppToService.feature (hunk 3 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature:62]: `a-d-s` stands for `app-databse-servicebinding` but now I replaced databse with `backend-service` so I changed it to `a-b-s`.

File: test/acceptance/features/bindAppToService.feature (hunk 5 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature:86]: Same explanation as above.

File: test/acceptance/features/bindAppToService.feature (hunk 4 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature:42]: Not every but most of the scenarios need secrets so instead of executing every time I added it here.

File: test/acceptance/features/bindAppToService.feature (hunk 3 above)
Comment (pmacik) [reply test/acceptance/features/bindAppToService.feature:62]: In the examples have permutations of App, DB and ServiceBinding - therefore `a-d-s`. However, here we have App, Service and the Binding, so what about using `s` instead of `d` and `b` instead of `s` so we would have permutations of `a-s-b`? 

I like the idea of naming things properly, consistently and uniquely to avoid collisions.

File: test/acceptance/features/bindAppToMultipleServices.feature (hunk 7 above)
Comment (pmacik) [reply test/acceptance/features/bindAppToMultipleServices.feature:10]: I agree with @pedjak, we should depend on CSV only if that is essential for the scenario itself

File: test/acceptance/features/bindAppToService.feature (hunk 4 above)
Comment (pedjak) [reply test/acceptance/features/bindAppToService.feature:42]: "Bacground" should contain steps that are needed by all scenarios.

File: test/acceptance/features/bindAppToMultipleServices.feature (hunk 7 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToMultipleServices.feature]: Okay, I removed `Operator X is running` step from various scenarios PTAL.

File: test/acceptance/features/bindAppToService.feature (hunk 8 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature]: Removed it.

File: test/acceptance/features/bindAppToService.feature (hunk 4 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature:43]: Done.

File: test/acceptance/features/bindAppToService.feature [hunk 9]
```diff
@@ -344,7 +473,7 @@ Feature: Bind an application to a service
             """
             true
             """
-
+    @olm
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:476]: does this scenario depends on the existence of a CSV? If not, we should not introduce this tag

File: test/acceptance/features/bindAppToServiceUsingSecret.feature [hunk 10]
```diff
@@ -10,8 +10,7 @@ Feature: Bind values from a secret referred in backing service resource
     @olm
     Scenario: Inject into app a key from a secret referred within service resource
         Binding definition is declared on service CRD.
-        Given OLM Operator "backend" is running
```
Comment (pedjak) [inline test/acceptance/features/bindAppToServiceUsingSecret.feature:13]: if removed, then we should remove `@olm` tag as well?

File: test/acceptance/features/bindAppToServiceUsingSecret.feature [hunk 11]
```diff
@@ -80,9 +79,7 @@ Feature: Bind values from a secret referred in backing service resource
 
     @olm
     Scenario: Inject into app all keys from a secret referred within service resource
-
-        Given OLM Operator "backend" is running
```
Comment (pedjak) [inline test/acceptance/features/bindAppToServiceUsingSecret.feature:84]: same question as above.

File: test/acceptance/features/bindWithCustomNamingStrategies.feature [hunk 12]
```diff
@@ -7,11 +7,29 @@ Feature: Bind an application to a service using custom naming strategies
...
+                    served: true
+                    storage: true
+                scope: Namespaced
+                names:
+                    plural: backends
+                    singular: backend
+                    kind: Backend
+                    shortNames:
+                      - bk
+            """
```
Comment (pedjak) [inline test/acceptance/features/bindWithCustomNamingStrategies.feature:29]: Let's follow DRY and avoid copying - wrap this into a step like `CustomResourceDefinition backends.stable.example.com is available`

File: test/acceptance/features/postgresqlOperatorExamples.feature [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+                    kind: Database
+                    name: db-demo-2
+            """
+        Then Service Binding "binding-request-1" is ready
+        And Service Binding "binding-request-2" is ready
+        And "nodejs-app" deployment must contain SBR name "binding-request-1"
+        And "nodejs-app" deployment must contain SBR name "binding-request-2"
+
+    # https://github.com/redhat-developer/service-binding-operator/tree/master/examples/pod_spec_path
+    Scenario: Specify container's path in Service Binding
```
Comment (pedjak) [inline test/acceptance/features/postgresqlOperatorExamples.feature:253]: do we have now the same scenario but without database available in some other feature file?

File: test/acceptance/features/postgresqlOperatorExamples.feature [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+                  - group: postgresql.baiju.dev
+                    version: v1alpha1
+                    kind: Database
+                    name: db-demo-csp
+                    id: zzz
+            """
+        Then Service Binding "binding-request-csp" is ready
+        And Secret "binding-request-csp" has been injected in to CR "demo-appconfig-csp" of kind "AppConfig" at path "{.spec.spec.containers[0].envFrom[0].secretRef.name}"
+
+    Scenario: Specify secret's path in Service Binding
```
Comment (pedjak) [inline test/acceptance/features/postgresqlOperatorExamples.feature:297]: same as above.

**Review by pedjak — CHANGES_REQUESTED** [review]

File: test/acceptance/features/bindAppToService.feature (hunk 9 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature:476]: I guess it depends on CSV, CSV file define some descriptors. It is using `OLM Operator "backend-new-spec" is running` step.

File: test/acceptance/features/bindAppToServiceUsingSecret.feature (hunk 10 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToServiceUsingSecret.feature:13]: Yes.

File: test/acceptance/features/bindAppToServiceUsingSecret.feature (hunk 11 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToServiceUsingSecret.feature:84]: Yes.

File: test/acceptance/features/postgresqlOperatorExamples.feature (hunk 13 above)
Comment (pratikjagrut) [reply test/acceptance/features/postgresqlOperatorExamples.feature:253]: Yes we've at https://github.com/redhat-developer/service-binding-operator/pull/880/files#diff-ca848b998fbd78a0989748822c09065858004a5e79b457ae9b57ad5ed9dcd25cR117

File: test/acceptance/features/postgresqlOperatorExamples.feature (hunk 14 above)
Comment (pratikjagrut) [reply test/acceptance/features/postgresqlOperatorExamples.feature:297]: Yes we've at https://github.com/redhat-developer/service-binding-operator/pull/880/files#diff-ca848b998fbd78a0989748822c09065858004a5e79b457ae9b57ad5ed9dcd25cR117

File: test/acceptance/features/bindWithCustomNamingStrategies.feature (hunk 12 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindWithCustomNamingStrategies.feature]: PTAL.

File: test/acceptance/features/bindAppToService.feature (hunk 9 above)
Comment (pedjak) [reply test/acceptance/features/bindAppToService.feature:457]: thanks for checking.

File: test/acceptance/features/steps/steps.py [hunk 15]
```diff
@@ -557,3 +557,30 @@ def update_binding_secret(context):
     output = openshift.apply(yaml.dump(secret_yaml), context.namespace.name)
     result = re.search(rf'.*{context.sb_secret}.*(created|unchanged|configured)', output)
     assert result is not None, f"Unable to apply YAML for binding secret '{context.sb_secret}': {output}"
+
+
+@given(u'CustomResourceDefinition backends.stable.example.com is available')
+def backends_crd(context):
```
Comment (pmacik) [inline test/acceptance/features/steps/steps.py:563]: Could you please move the CRD YAML into the separate file under `test/acceptance/resources` and load it here from it, so it is not hard-coded?

Comment (pmacik) [reply test/acceptance/features/steps/steps.py:563]: I can see that the actual CRD is already there as part of `test/acceptance/resources/backend-new-spec.operator.manifest.yaml`.

File: test/acceptance/features/postgresqlOperatorExamples.feature [hunk 16]
```diff
@@ -0,0 +1,331 @@
+@examples
```
Comment (pedjak) [inline test/acceptance/features/postgresqlOperatorExamples.feature:1]: are we sure that `@examples` scenarios are passing? How about that we create additional job that can run examples on demand/manually? https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/ That should be a part of the current PR.

File: test/acceptance/features/steps/steps.py [hunk 17]
```diff
@@ -348,7 +348,12 @@ def operator_manifest_installed(context, backend_service):
         ns = context.namespace.name
     else:
         ns = None
-    _ = openshift.apply_yaml_file(os.path.join(os.getcwd(), "test/acceptance/resources/", backend_service + ".operator.manifest.yaml"), namespace=ns)
+
+    if backend_service == "backend":
+        _ = openshift.apply_yaml_file(os.path.join(os.getcwd(), "test/acceptance/resources/", backend_service + ".operator.manifest.yaml"), namespace=ns)
+        _ = openshift.apply_yaml_file(os.path.join(os.getcwd(), "test/acceptance/resources/backend_crd.yaml"), namespace=ns)
+    else:
+        _ = openshift.apply_yaml_file(os.path.join(os.getcwd(), "test/acceptance/resources/", backend_service + ".operator.manifest.yaml"), namespace=ns)
```
Comment (pedjak) [inline test/acceptance/features/steps/steps.py:356]: let keep things simple - each fake operator is defined by single yaml file, i.e. all resources should be there, and not distributed across various files. In this way, the stop logic can be reverted to the previous state.

File: test/acceptance/features/steps/steps.py [hunk 18]
```diff
@@ -557,3 +562,13 @@ def update_binding_secret(context):
...
+
+
+@given(u'CustomResourceDefinition backends.stable.example.com is available')
+def backends_crd(context):
+    openshift = Openshift()
+    if "namespace" in context:
+        ns = context.namespace.name
+    else:
+        ns = None
+    _ = openshift.apply_yaml_file(os.path.join(os.getcwd(), "test/acceptance/resources/backend_crd.yaml"), namespace=ns)
```
Comment (pedjak) [inline test/acceptance/features/steps/steps.py:574]: please avoid DRY - this looks very similar to `operator_manifest_installed` method, let's generalize it if needed, and add additional step definition to it.

File: .github/workflows/examples_manually_trigger.yaml [hunk 19]
```diff
@@ -0,0 +1,46 @@
+name: Run Examples
```
Comment (pedjak) [inline .github/workflows/examples_manually_trigger.yaml:1]: is there any run of this job somewhere, so we can be sure it is green before merging this PR?

File: test/acceptance/features/steps/steps.py (hunk 17 above)
Comment (pratikjagrut) [reply test/acceptance/features/steps/steps.py]: Done.

File: test/acceptance/features/steps/steps.py (hunk 18 above)
Comment (pratikjagrut) [reply test/acceptance/features/steps/steps.py]: Done.

File: .github/workflows/examples_manually_trigger.yaml (hunk 19 above)
Comment (pratikjagrut) [reply .github/workflows/examples_manually_trigger.yaml:1]: I didn't find any way here, so I loaded this new job on my fork's master and I'll update the link once everything is fine.

File: .github/workflows/examples_manually_trigger.yaml [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+on: workflow_dispatch
+
+env:
+  GO111MODULE: on
+  SDK_VERSION: "1.3.0"
+  OPM_VERSION: "1.15.2"
```
Comment (Avni-Sharma) [inline .github/workflows/examples_manually_trigger.yaml:8]: The target `release-operator` does not require opm . The target `prepare-operatorhub-pr` requires opm binary

Comment (pmacik) [reply .github/workflows/examples_manually_trigger.yaml:8]: @Avni-Sharma `release-operator` actually does depend on `opm` transitively from [`index-image`](https://github.com/redhat-developer/service-binding-operator/blob/master/Makefile#L252) through `release-operator`->`push-index-image`->`index-image`

File: .github/workflows/examples_manually_trigger.yaml (hunk 19 above)
Comment (pratikjagrut) [reply .github/workflows/examples_manually_trigger.yaml:1]: @pedjak The test run of this workflow https://github.com/pratikjagrut/service-binding-operator/actions/workflows/examples_manually_trigger.yaml

**Review by pedjak — APPROVED** [review]
please squash

**Comment by pratikjagrut** [comment]
> please squash

Squashed.

**Comment by pratikjagrut** [comment]
Rebased.
